### PR TITLE
feat(ext/net): extract TLS key and certificate from interfaces

### PR DESCRIPTION
**Pull request by bartlomieju**
Relands #23325

### CODE REVIEW

**Comment by iuioiua** [comment]
Previously, in 1.41, `Deno.listenTls({ port: 443 })` would throw with:
```
Uncaught (in promise) Error: `cert` is not specified.
```
Now, this same snippet causes a panic because `loadTlsKeyPair()` returns `op_tls_key_null()`. Perhaps, this should be addressed in this PR.

**Comment by iuioiua** [comment]
To add to my previous comment, perhaps we should just error within `loadTlsKeyPair()` instead of returning `op_tls_key_null()`.

**Comment by mmastrac** [comment]
> To add to my previous comment, perhaps we should just error within `loadTlsKeyPair()` instead of returning `op_tls_key_null()`.

We need to have an option to be successful if no key is specified for connectTls, however (key is optional there).

**Comment by mmastrac** [comment]
> Previously, in 1.41, `Deno.listenTls({ port: 443 })` would throw with:
> 
> ```
> Uncaught (in promise) Error: `cert` is not specified.
> ```
> 
> Now, this same snippet causes a panic because `loadTlsKeyPair()` returns `op_tls_key_null()`. Perhaps this should be addressed in this PR. Note: this is what's currently blocking #23271 and #23270.

I'm not sure if this is true -- I will add a fix around it, but I'm seeing it correctly detect that a certificate and/or key is missing:

```
> Deno.listenTls()
Uncaught TypeError: Cannot destructure property 'port' of 'undefined' as it is undefined.
    at Object.listenTls (ext:deno_net/02_tls.js:230:3)
    at <anonymous>:1:27
> Deno.listenTls({})
Uncaught TypeError: A key and certificate are required for `listenTls`
    at Object.listenTls (ext:deno_net/02_tls.js:241:11)
    at <anonymous>:1:27
> Deno.listenTls({ cert: "" })
Uncaught TypeError: If `cert` is specified, `key` must be specified as well for `Deno.listenTls`.
    at both (ext:deno_net/02_tls.js:177:13)
    at loadTlsKeyPair (ext:deno_net/02_tls.js:189:3)
    at Object.listenTls (ext:deno_net/02_tls.js:243:19)
    at <anonymous>:1:27
```

File: cli/tsc/dts/lib.deno.ns.d.ts [hunk 1]
```diff
@@ -6490,7 +6484,10 @@ declare namespace Deno {
    * @category HTTP Server
    */
   export function serve(
-    options: ServeOptions | ServeTlsOptions,
+    options:
+      | ServeOptions
+      | ServeTlsOptions
+      | (ServeTlsOptions & TlsCertifiedKeyOptions),
     handler: ServeHandler,
```
Comment (mmastrac) [inline cli/tsc/dts/lib.deno.ns.d.ts:6491]: @marvinhagemeister @bartlomieju What do you guys think of this? We can make `ServeTlsOptions` an empty interface which will allow for older code using `ServeTlsOptions` to continue to typecheck.

**Comment by iuioiua** [comment]
> Deno.listenTls({ port: 443 })

To clarify, I saw the error for `Deno.listenTls({ port: 443 })` specifically. I can confirm this is the case for the latest stable.

**Review by marvinhagemeister — APPROVED** [review]

**Pull request merged**